### DISCUSSION TITLE
How to connect multiple SPI Devices using MicroPython?

### DISCUSSION THREAD

**Discussion by neliel123** [micropython/micropython#13371] [Hardware & Peripherals]
Hi,
How do you connect two SPI devices in one microcontroller using MicroPython?
I could not find an example (or maybe I don't know the correct keyword).

I have a Raspberry Pi Pico W with a TFT  Display and MFRC522 RFID reader.  They both used  SPI.  How do you connect those two in my Pico and how can I program them using MicroPython?

Still learning about MicroPython

**Comment by robert-hh** [comment]
You can connect them both at the same SPI bus with CLK, MOSI, MISO. Only the CS pins must be different. You use the same SPI object for both devices and use the CS pins to address them. If they run at different speeds, you can use spi.init(baudrate=xxx) to change the speed as requested.

**Reply by mcauser** [reply]
For the visual learners:
```
mcu       device1   device2   device3
mosi ---- mosi ---- mosi ---- mosi
miso ---- miso ---- miso ---- miso
clk ----- clk ----- clk ----- clk
cs1 ----- cs    .-- cs    .-- cs
cs2 -----------`         /
cs3 --------------------`
```

**Comment by GC-RnD** [comment]
I’m using MicroPython on an ESP32‑C6 and trying to connect multiple SPI devices 
(e.g., display, touch controller, SD card, radio module, etc.) on the single general‑purpose SPI bus that the C6 hardware provides.

I found this statement in PR #16702:
“C3 and C6 only have one available SPI bus, so SDCard and machine.SPI cannot be used together.”

This implies that SDCard() takes exclusive control of the SPI bus, 
preventing any additional SPI devices from being used at the same time.

My questions for the MicroPython maintainers and community:
Is this limitation expected to be temporary or permanent for the ESP32‑C6 port?
Is SPI bus sharing (multiple devices on the same bus with separate CS lines) planned for the C6?
Is there any ongoing work, discussion, or roadmap item related to enabling SPI sharing on the C6?
Is the limitation due to missing features in the MicroPython port (e.g., DMA, bus locking, transaction queues), or is it an ESP‑IDF constraint?

The ESP32‑C6 hardware and ESP‑IDF both support multiple SPI devices on the same bus, <-----------
so I’m trying to understand whether MicroPython intends to support this as well.

Any information, guidance, or plans regarding SPI sharing on the ESP32‑C6 would be greatly appreciated.

**Reply by peterhinch** [reply]
It is normally possible to share an SPI bus, so long as every device properly supports `CS/`. I do this with touch displays, where some hosts have only one bus. You deassert `CS/` for all unused devices and assert it for the one you intend to access. In my case the touch controller requires a much lower baudrate than the display, so the arbitration needs to address this as well as `CS/`.

I've not used the C6 but I can't see an obvious reason why it shouldn't work so long as you code the necessary arbitration.

**Reply by scruss** [reply]
It looks like using SDCard in [width=1](https://github.com/micropython/micropython/issues/16526#issuecomment-2574977032) (SPI) mode allows you to share the bus with other SPI devices. Specify a different CS for the devices, though

**Comment by WangilaWakhongola** [comment]
To connect multiple SPI devices in MicroPython, use separate CS (Chip Select) pins for each device while sharing the same SPI bus (MOSI, MISO, SCK).
pythonfrom machine import SPI, Pin

spi = SPI(1, baudrate=1000000, polarity=0, phase=0,
          sck=Pin(14), mosi=Pin(13), miso=Pin(12))

cs1 = Pin(10, Pin.OUT)
cs2 = Pin(11, Pin.OUT)

#setting
cs1.value(1)
cs2.value(1)

def spi_transfer(cs, data):
    cs.value(0)          
    result = spi.read(len(data))
    cs.value(1)          
    return result

Key rules:
  1.  Only pull one CS pin LOW at a time
  2.  Keep all other CS pins HIGH while communicating
  3.  All devices share the same SPI, MOSI, MISO, SCK pins
  4.  Each device gets its own unique CS pin

**Reply by Josverl** [reply]
Tip: use 'triple backtick'
` ```py `
`# code goes here `
` ``` ` 
around your code snippets to get syntax highlighting and proper indenting.

**Comment by WangilaWakhongola** [comment]
from machine import SPI, Pin

spi = SPI(1, baudrate=1000000, sck=Pin(14), mosi=Pin(13), miso=Pin(12))

cs1 = Pin(10, Pin.OUT, value=1)
cs2 = Pin(11, Pin.OUT, value=1)

# To communicate with device 1
cs1.value(0)
spi.write(b'\x1')
cs1.value(1)
